aws_lambda_powertools/event_handler/api_gateway.py

```diff
@@ -538,19 +597,74 @@ def clear_context(self):
         self.context.clear()
 
 
-class MiddlewareHandler:
-    """Represents a callable Middleware function used to build a middleware stack when a route is resolved"""
+class MiddlewareStackWrapper:
+    """
+    creates a Middle Stack Wrapper instance to be used as a "Frame" in the overall stack of
+    middleware functions.  Each instance contains the current middleware to be executed and the next
+    middleware function to be called in the stack.
+
+    In this way the middleware stack is constructed in a recursive fashion, with each middleware
+    calling the next as a simple function call.  The actual Python call-stack will contain
+    each MiddlewareStackWrapper "Frame", meaning any Middleware function can cause the
+    entire Middleware call chain to be exited early (short-circuited) by raising an exception
+    or by simply returning early with a custom Response.  The decision to short-circuit the middleware
+    chain is at the user's discretion but instantly available due to the Wrapped nature of the
+    callable constructs in the Middleware stack and each Middleware function having complete control over
+    whether the "Next" handler in the stack is called or not.
+
+    Parameters
+    ----------
+    handler : Callable
+        The API route handler to be invoked when this class is executed.
+    next_middleware : Callable
+        The next middleware in the middleware stack.
+    """
 
     def __init__(
         self,
-        handler: Callable[..., Any],
-        next_handler: Callable[..., Any],
+        handler: Callable,
+        next_middleware: Callable,
     ) -> None:
-        self.handler: Callable[..., Any] = handler
-        self.next_handler: Callable[..., Any] = next_handler
+        self.handler: Callable = handler
+        self.next_middleware: Callable = next_middleware
 
-    def __call__(self, app: BaseRouter, **kwargs) -> Any:
-        return self.handler(app, self.next_handler, **kwargs)
+    def __call__(self, app: BaseRouter, **kwargs) -> Union[Dict, Tuple, Response]:
+        """
+        Call the middleware stack wrapper
+
+        Parameters
+        ----------
+        app: BaseRouter
+            The router instance
+        **kwargs
+            Any additional arguments to pass to the middleware stack wrapper
+
+        Returns
+        -------
+        Union[Dict, Tuple, Response]
+            (tech-debt for backward compatability).  The response type should be a
+            Response object in all cases excepting when the oiginal API rout handler
+            is executed which will return one of 3 outputs.
+
+        """
+        return self.handler(app, self.next_middleware, **kwargs)
+
+
+# No type checking possible due to Dependency order of definition ()
+def registered_api_adapter(app, get_response: Callable[..., Any], **kwargs) -> Union[Dict, Tuple, Response]:
```

Comment:
```suggestion
def registered_api_adapter(app: "ApiGatewayResolver", get_response: Callable[..., Any], **kwargs) -> Union[Dict, Tuple, Response]:
```

You can use forward reference type annotation under these circumstances. The string will be resolved to its type if available

Reply:
Maybe we could have a better name, can't think of anything now.

Also, can we make this private or do we need to expose to customers?